### ISSUE TITLE
To Next Level (TNL) Algorithm Modifications

### ISSUE THREAD

**Issue by PhilipHow** [HabitRPG/habitica#458]
Hey!

After looking at some of the Wiki pages I noticed the TNL algorithm could use some discussion.

It's mentioned on the [Algorithms](https://github.com/lefnire/habitrpg/wiki/Algorithms) page but the original JSFiddle for the [StackExchange](http://gamedev.stackexchange.com/questions/13638/algorithm-for-dynamically-calculating-a-level-based-on-experience-points) discussion were out and producing some funny numbers.

The square relation (from now on known as Square) formula seems to provide a realistic increase in difficulty, it goes like this:

![CodeCogsEqn](https://f.cloud.github.com/assets/1628871/176236/4cff1fbc-7b6e-11e2-84d1-d76424d1185b.gif)

The second part of the equation lowers the 1st level requirement to zero.

Here's a sexy Excel graph with the level curves using some suggested values for the Constant, the levels go up to 100:

![graph](https://f.cloud.github.com/assets/1628871/176261/c7fb1c2a-7b6e-11e2-9c0f-9f587b9377ca.png)

And you can play with it yourself on [JSFiddle](http://jsfiddle.net/GxQb8/) (it was running incredibly slow for me however)

As you can see the Tibia algorithm spikes super high, Square 10 exceed it around the level 10 mark before it becomes more reasonable. Square 5 seems to be decent model to use, it hangs around the Linear Algorithm until about level 5 where the differences become noticable.

The SE discussion mentioned that Square Relation was good for RPGs where your higher level doesn't help increase your experience gathering skills. Although this is sort of the case within HRPG, purchasing items can increase your XP gathering rate. This discussion also links with some of the existing discussions around game balancing - some of the planned features would probably mess with this more.

A modified version of the Tibia algorithm would probably be best suited to this, something that recognise a slight increase in XP gathering ability, but doesn't go crazy like the Tibia currently does.

A good way to deduce this would be to model XP gain over time, for differing play styles and taking into account some of the game mechanics that increase your XP significantly. You could then enter in how long you wanted each type of player to progress for each level, and set the XP requirement algorithm to match this!

Whew, this was longer than I anticipated, but hopefully we can start some good discussion!

**Comment by lefnire** [comment]
This is great! I'll merge this in later if nobody has any concerns

**Comment by PhilipHow** [comment]
So I created three (very basic) models to view level/day, assuming 5/15/30 completed tasks per day for low, medium and high active users. The model assumes you will always buy the most expensive XP enhancing weapon as soon as you can. You get a base of 10XP for each completed task.

Here are the curves for the Square(4) algorithm:
![4](https://f.cloud.github.com/assets/1628871/187471/82b53122-7d5e-11e2-9663-7ba88118beaf.png)

And here's Square(5)
![5](https://f.cloud.github.com/assets/1628871/187472/857e32e6-7d5e-11e2-9b67-229ca5ca0142.png)

Damn, they need axes (ahem), X is days, Y is level.

This is just for the first 30 days, I can add more if requested.

Basically @lefnire, what kind of progression did you have in mind? Are there any planned weapons (i.e. backer weapons) that would affect this? The model also uses the 0.03 adjustments for both XP and Gold per level + weapon increase, let me know if you plan to fiddle with these!

So I had clearly misunderstood previously, a player's ability to gain XP **does** increase by level, which means you'll get a pretty linear progression like you see in the above graphs. Is this intentional, or should there be a slight curve in the progression graph?

**Comment by lefnire** [comment]
These are great. FYI, will tackle this soon as I get through the critical bugs & features - so maybe 3 weeks from now (unless someone does a PR against my algorithms before I get there)

**Comment by Pandoro** [comment]
I will close this considering @MrConcept has done the pull request which has been merged already regarding the new stuff. If this is actually something else :x Please reopen it again.
